common.rkt

```diff
@@ -146,13 +146,13 @@
                                (state (extend-sub t (reified-index index) (state-sub st)) (state-diseq st) (state-types st)))
                     (else      st)))))
     (let* ((walked-sub (walk* tm results))
-          (diseq-null (null? (state-diseq st)))
-          (types-null (null? (state-types st)))
-          (types (walk* (map pretty-types (state-types st)) results))
-          (diseq (if diseq-null '() (list (walk* (cons '=/= (map pretty-diseq (state-diseq st))) results)))))
-      (if (and diseq-null types-null)
+           (types (walk* (map pretty-types (state-types st)) results))
+           (diseq (map pretty-diseq (state-diseq st)))
+           (diseq (if (null? diseq) '() (list (walk* (cons '=/= diseq) results))))
+           (cxs (append types diseq)))
+      (if (null? cxs)
           walked-sub
-          (Ans walked-sub (append types diseq))))))
+          (Ans walked-sub cxs)))))
```

Comment:
Let's simplify the reified results some more.

I noticed you don't have any tests involving constraints on `fresh` variables.  You should add some, which should point out that some of your reified results might still contain raw variables, even after `walk*`ing with `results`.

Once you're reifying an answer, you know all of your constraints are satisfiable.  But the situation might be even better than that.  You may be able to discard some of the constraints, providing a simpler answer to the user.  There are still two forms of constraints left for you to remove: subsumed disequality constraints, and irrelevant variable constraints.

I described how to handle disequality subsumption in one of the earlier review comments: you can sort the diseqs by length, then reapply them in shortest-to-longest order.  This way, some of the disequalities applied earlier may cause later ones to disappear, even if they didn't during normal execution.

Irrelevant variables are those which don't appear in the answer term.  For instance, in your implementation, the query:

```
(run* (a b) (fresh (c) (== a b) (=/= c (cons a b))))
```

will produce an answer that retains a disequality constraint on `c`.  However, the answer term will only mention `a` and `b` (and indicate that they are the same in this case).  Because there is no value of `a` (and `b`, since they're equal) that can invalidate the disequality constraint on `c`, that disequality constraint is vacuously true: its presence does not add any information.  You can throw it away.

Here is another example, where the fresh variable is on the other side of the disequality:

```
(run* (a b) (fresh (c) (== a b) (=/= a (cons c 5))))
```

The disequality constraint mentioning `c` here will also be vacuously true.  Regardless of what `a` is, you can find a `c` such that `a` will not be equal to a pair of `c` and `5`.

This also goes for more complex disequality constraints:

```
(run* (a b) (fresh (c) (=/= (cons c 5) (cons a b))))
```

Here, you'll end up with a disequality constraint claiming the following: `(OR (=/= c a) (=/= 5 b))`

Even though it looks like `b` is constrained independently of `c`, you can throw the entire disequality disjunction away.  The reason is, we can always find some irrelevant `c` that is not equal to `a`.  And if one branch of a disjunction is vacuously true, the entire disjunction is.


Same story for type constraints:

```
(run* (a b) (fresh (c) (== a b) (numbero c)))
```

`c` doesn't participate in the answer term, so we can always find some `c` that is a number.  We can throw that type constraint away.